ocaml/testsuite/tests/typing-modes/module.ml

```diff
@@ -0,0 +1,88 @@
+(* TEST
```

Comment:
I think there should also be some tests around functor application: functor application can cause mutation, and applying a functor in a function (even without first-class modules) should cause the function to close over both the functor and the argument, causing modes to change.

Reply:
I agree with what you said, but think it's better to decompose:
- Functor application can cause mutation, this is reflected by the fact that functors (like all modules) are `nonportable`.
- Closing over `F(M)` will constrain modes. But really from the perspective of the implementation, we are closing over two modules `F` and `M`. 

In summary, the logic needed is already there for regular modules. That being said, I will still add the tests as integration tests.

Reply:
Well in fact I suggest we don't test this at all. If the test involves `F(X)`, we can't tell if we are testing `F` or `X`. If the test is about `F` alone, then it doesn't really seem like a new test to me. That is, `F` alone is obviously treated the same as `X` alone.

Reply:
There's something not quite right here. Suppose I have

```ocaml
let mylength list = List.length list
```

Does this close over `List`? If so, then because all modules are nonportable, `mylength` should be nonportable, too. Yet I don't think that's the case -- nor should it be. When we're projecting out of a module (as we almost always do), we can just look at the projection's modality (soon, we'll allow `List.length` to be portable). So, how do we have a function close over a module? We could use a first-class module to capture an entire module... but I don't want to test first-class modules here. It seems to me the only other way is to do a functor application, with `let module M = F(X) in ...`. It's true that this also says something about `X`, but I don't see a way around that.

My bottom line: something is different about functor application than just projecting from a module, because functor application is actually effectful. And I think that capturing that effect should somehow be reflected in our tests.

Reply:
Good question. @lpw25 and I had a previous discussion about this, and think `mylength` doesn't close over `List`. See the section "Closing over values in modules" in the internal document "modes and modules".

Reply:
Right. I agree. But then there must be some test that can witness side effects that happen during functor application. This is why I wanted a test looking at `F(X)`, because I think that having a functor application in the middle of a function might affect modes differently than projecting values.

Reply:
The fact that `F(X)` might cause mutation is modelled by `F` being nonportable. Functor application is not essential here: a function that captures this `F` without applying it will still be `nonportable`.

And it's same logic (in #2431 ) that handles closing over modules, whether it's a structure or a functor. Tests for the former already exists; I don't mind adding the latter just to be sure.

Reply:
Yes, I agree that closing over `F` is the real issue. The problem is that the test that closes over `F` requires using a first-class module, which has its own set of mode restrictions, etc. My desire for functor application was a way of closing over `F` without invoking first-class modules.

Reply:
The problem with `F(X)` is that this also closes over `X`, which is treated similarly to `F`. So even if the closure's mode is forced as expected, we don't know if it's because of `F` or `X`. (Hence my second comment above).

Reply:
I continue to think this test is worthwhile. I agree that we can't test exactly the thing we'd like -- closing over `F` without first-class modules or any other modules -- but a test that covers too much (both `F` and `X`) is better than none at all. And `X` can just be `struct end`! Or not even that: if `F` is a generative functor, we don't even need the `X`. (Though actually I think we should test both generative and applicative functors.)

Reply:
I suppose I can define `X` inside the closure, then it's not closed over `X`.